src/server/game/Achievements/AchievementMgr.cpp

```diff
@@ -511,8 +511,8 @@ void AchievementMgr::ResetAchievementCriteria(AchievementCriteriaCondition condi
 {
     TC_LOG_DEBUG("achievement", "AchievementMgr::ResetAchievementCriteria(%u, %u, %u)", condition, value, evenIfCriteriaComplete);
 
-    // disable for gamemasters with GM-mode enabled
-    if (m_player->IsGameMaster())
+    // disable for gamemasters with GM-mode enabled or when achievements for GM are disabled
+    if (m_player->IsGameMaster() || (m_player->CanBeGameMaster() && !sWorld->getBoolConfig(CONFIG_ALLOW_GM_ACHIEVEMENTS)))
```

Comment:
I would prefer a new method in AchievementMgr accepting a Player* as parameter and returning bool as a wrapper around these checks since they are repeated already in 3 different places